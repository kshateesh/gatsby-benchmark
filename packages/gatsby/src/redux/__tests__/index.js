const _ = require(`lodash`)
const path = require(`path`)

const writeToCache = jest.spyOn(require(`../persist`), `writeToCache`)
const { saveState, store, readState } = require(`../index`)

const {
  actions: { createPage },
} = require(`../actions`)

const mockWrittenContent = new Map()
const mockCompatiblePath = path
jest.mock(`fs-extra`, () => {
  return {
    writeFileSync: jest.fn((file, content) =>
      mockWrittenContent.set(file, content)
    ),
    readFileSync: jest.fn(file => mockWrittenContent.get(file)),
    moveSync: jest.fn((from, to) => {
      // This will only work for folders if they are always the full prefix
      // of the file... (that goes for both input dirs). That's the case here.
      if (mockWrittenContent.has(to)) {
        throw new Error(`File/folder exists`)
      }

      // Move all files in this folder as well ... :/
      mockWrittenContent.forEach((value, key) => {
        if (key.startsWith(from)) {
          // rename('foo/bar', 'a/b/c') => foo/bar/ding.js -> a/b/c/ding.js
          // (.replace with string arg will only replace the first occurrence)
          mockWrittenContent.set(
            key.replace(from, to),
            mockWrittenContent.get(key)
          )
          mockWrittenContent.delete(key)
        }
      })
    }),
    existsSync: jest.fn(target => mockWrittenContent.has(target)),
    mkdtempSync: jest.fn(suffix => {
      const dir = mockCompatiblePath.join(
        `some`,
        `tmp` + suffix + Math.random()
      )
      mockWrittenContent.set(dir, Buffer(`empty dir`))
      return dir
    }),
    removeSync: jest.fn(file => mockWrittenContent.delete(file)),
  }
})
jest.mock(`glob`, () => {
  return {
    sync: jest.fn(pattern => {
      // Tricky.
      // Expecting a path prefix, ending with star. Else this won't work :/
      if (pattern.slice(-1) !== `*`) {
        throw new Error(`Expected pattern ending with star`)
      }
      const globPrefix = pattern.slice(0, -1)
      if (globPrefix.includes(`*`)) {
        throw new Error(`Expected pattern to be a prefix`)
      }
      const files = []
      mockWrittenContent.forEach((value, key) => {
        if (key.startsWith(globPrefix)) {
          files.push(key)
        }
      })
      return files
    }),
  }
})

function getFakeNodes() {
  // Set nodes to something or the cache will fail because it asserts this
  // Actual nodes content should match TS type; these are verified
  const map /* : Map<string, IReduxNode>*/ = new Map()
  map.set(`pageA`, {
    id: `pageA`,
    internal: {
      type: `Ding`,
    },
  })
  map.set(`pageB`, {
    id: `pageB`,
    internal: {
      type: `Dong`,
    },
  })
  return map
}

describe(`redux db`, () => {
  const initialComponentsState = _.cloneDeep(store.getState().components)

  beforeEach(() => {
    store.dispatch(
      createPage(
        {
          path: `/my-sweet-new-page/`,
          // seems like jest serializer doesn't play nice with Maps on Windows
          component: `/Users/username/dev/site/src/templates/my-sweet-new-page.js`,
          // The context is passed as props to the component as well
          // as into the component's GraphQL query.
          context: {
            id: `123456`,
          },
        },
        { name: `default-site-plugin` }
      )
    )

    writeToCache.mockClear()
    mockWrittenContent.clear()
  })

  it(`should write redux cache to disk`, async () => {
    expect(initialComponentsState).toEqual(new Map())

    store.getState().nodes = getFakeNodes()

    await saveState()

    expect(writeToCache).toBeCalled()

    // reset state in memory
    store.dispatch({
      type: `DELETE_CACHE`,
    })
    // make sure store in memory is empty
    expect(store.getState().components).toEqual(initialComponentsState)

    // read data that was previously cached
    const data = readState()

    // make sure data was read and is not the same as our clean redux state
    expect(data.components).not.toEqual(initialComponentsState)

    expect(data).toMatchSnapshot()
  })

  describe(`GATSBY_DISABLE_CACHE_PERSISTENCE`, () => {
    beforeAll(() => {
      process.env.GATSBY_DISABLE_CACHE_PERSISTENCE = `truthy`
    })

    afterAll(() => {
      delete process.env.GATSBY_DISABLE_CACHE_PERSISTENCE
    })
    it(`shouldn't write redux cache to disk when GATSBY_DISABLE_CACHE_PERSISTENCE env var is used`, async () => {
      expect(initialComponentsState).toEqual(new Map())

      store.getState().nodes = getFakeNodes()

      await saveState()

      expect(writeToCache).not.toBeCalled()
    })
  })
})
