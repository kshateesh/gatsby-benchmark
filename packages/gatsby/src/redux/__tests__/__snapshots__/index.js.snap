// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux db should write redux cache to disk 1`] = `
Object {
  "components": Map {
    "/Users/username/dev/site/src/templates/my-sweet-new-page.js" => Object {
      "componentPath": "/Users/username/dev/site/src/templates/my-sweet-new-page.js",
      "isInBootstrap": true,
      "pages": Set {
        "/my-sweet-new-page/",
      },
      "query": "",
    },
  },
  "html": Object {
    "browserCompilationHash": "",
    "ssrCompilationHash": "",
    "trackedHtmlFiles": Map {
      "/my-sweet-new-page/" => Object {
        "dirty": 2,
        "isDeleted": false,
        "pageDataHash": "",
      },
    },
    "trackedStaticQueryResults": Map {},
  },
  "jobsV2": Object {
    "complete": Map {},
    "incomplete": Map {},
  },
  "nodes": Map {
    "pageA" => Object {
      "id": "pageA",
      "internal": Object {
        "type": "Ding",
      },
    },
    "pageB" => Object {
      "id": "pageB",
      "internal": Object {
        "type": "Dong",
      },
    },
  },
  "nodesByType": Map {
    "Ding" => Map {
      "pageA" => Object {
        "id": "pageA",
        "internal": Object {
          "type": "Ding",
        },
      },
    },
    "Dong" => Map {
      "pageB" => Object {
        "id": "pageB",
        "internal": Object {
          "type": "Dong",
        },
      },
    },
  },
  "pageDataStats": Map {},
  "pendingPageDataWrites": Object {
    "pagePaths": Set {},
  },
  "queries": Object {
    "byConnection": Map {},
    "byNode": Map {},
    "deletedQueries": Set {},
    "dirtyQueriesListToEmitViaWebsocket": Array [],
    "queryNodes": Map {},
    "trackedComponents": Map {
      "/Users/username/dev/site/src/templates/my-sweet-new-page.js" => Object {
        "componentPath": "/Users/username/dev/site/src/templates/my-sweet-new-page.js",
        "errors": 0,
        "pages": Set {
          "/my-sweet-new-page/",
        },
        "query": "",
      },
    },
    "trackedQueries": Map {
      "/my-sweet-new-page/" => Object {
        "dirty": 1,
        "running": 0,
      },
    },
  },
  "staticQueriesByTemplate": Map {},
  "staticQueryComponents": Map {},
  "status": Object {
    "PLUGINS_HASH": "",
    "plugins": Object {},
  },
  "webpackCompilationHash": "",
}
`;
